### DISCUSSION TITLE
How to create range datatype in SLQlchemy

### DISCUSSION THREAD

**Discussion by TanelS** [sqlalchemy/sqlalchemy#11207] [Usage Questions]
Hi

I'm using PostgreSQL (15.5) and SQLAlchemy 2.0.29, DB driver is psycopg (aka psycopg3). In PG I created new RANGE type `timerange`:

`CREATE TYPE timerange AS RANGE (SUBTYPE = time);`

As PG does not have time range type, I had to create one as I need that datatype. Yes, it is possible to use `TSRANGE` and ignore the date part but that is not elegant and involves inserting useless data. Time without timezone is fine, I'm not worried about turn-of-the-day issue.

Using DB console everything works, I can insert with plain SQL timeranges like `[13:30:00,14:05:00)`, also multiranges as new type has `timemultirange` type was created automatically.

I tried to use `UserDefinedType`:

```python
import sqlalchemy.types as types

class TimeRange(types.UserDefinedType):
    cache_ok = True

    def get_col_spec(self, **kw: Any) -> str:
        return "timerange"
```
Imported that class into my declarative model, hacked Alembic migration file and needed column appeared with proper type.
  
I am kind of able to inset data

```python
def add_times(time_from, time_to):
    with SessionLocal() as session:
        dates = ScheduleModel(times=(time_from, time_to))
        session.add(dates)
        session.commit()

if __name__ == '__main__':
	add_times('13:30', '14:05') 
```  

Result is written into DB with  only`'()'` bounds and I just don't know how to change custom type class so, that it could accept `Range` object like `Range(time_from, time_to, bounds='[]')`.  I left out `bind_processor` from the class, maybe this is the place to enable bounds, provided that class has something like `def __init__(self, bounds=None, **kw)`.

I'm bit new to SQLAlchemy custom types and would appreciate some guidance. 

Thank you!

**Comment by CaselIT** [comment]
Hi,

I think your best option here is to subclass one of the internal ranges defined here:
https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/postgresql/ranges.py

something like
```py
class TIMERANGE(AbstractSingleRange[time]):
    """Represent the custom PostgreSQL TIMERANGE type."""


@compiles(TIMERANGE)
def compile_varchar(element, compiler, **kw):
  return 'TIMERANGE'
```

**Reply by TanelS** [reply]
@CaselIT thanks! I will look into subclassing.

**Reply by TanelS** [reply]
@CaselIT your idea almost works.

```python
class TIMERANGE(AbstractSingleRange[time]):
    """Represent the custom PostgreSQL TIMERANGE type."""
    __visit_name__ = "TIMERANGE"


@compiles(TIMERANGE)
def compile_varchar(element, compiler, **kw):
    return 'TIMERANGE'

```    
 but if only I add  `__visit_name__ = "TIMERANGE"`  and  have 
 
 ```python
def visit_TIMERANGE(self, type_, **kw):
	return "TIMERANGE"
```
in `sqlalchemy.dialects.postgresql.base.PGTypeCompiler` class. Maybe I am overlooking something obvious ...

**Reply by CaselIT** [reply]
Try doing `@compiles(TIMERANGE, 'postgresql')` there.
I think that `compiles` should be able to override the default type compiler of a dialect, even for existing types, so It would work in this case.

If it does not, report back that I'll try looking what's up

**Reply by TanelS** [reply]
Tried compile part with the dialect already as it seemed logical. Did not help. I have a suspicion that exception is risen right after check if method `visit_<new_type>()` is is not in the class `PGTypeCompiler`.  I wish there were a way to patch necessary classes with new methods. One can inherit, but then I cannot use the new class to be part of the pipeline in special cases.

**Comment by CaselIT** [comment]
This seems like a bug, cc @zzzeek 

```py
from datetime import time

import sqlalchemy as sa
from sqlalchemy import orm
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.dialects import postgresql as pg


class TIMERANGE(pg.AbstractSingleRange[time]):
    """Represent the custom PostgreSQL TIMERANGE type."""

    __visit_name__ = "TIMERANGE"


@compiles(TIMERANGE, "postgresql")
def compile_timerange(element, compiler, **kw):
    print(">>", type(element))
    return "TIMERANGE"


e = sa.create_engine("postgresql+psycopg://scott:tiger@127.0.0.1:5432/test")
e.connect().close()


class Base(orm.DeclarativeBase):
    pass


class Mod(Base):
    __tablename__ = "mod"
    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)
    rr: orm.Mapped[pg.Range[time]] = orm.mapped_column(TIMERANGE)


e.echo = True
with e.connect() as c:
    c.exec_driver_sql("CREATE TYPE timerange AS RANGE (SUBTYPE = time)")
    Base.metadata.create_all(c)

    with orm.Session(c) as s:
        s.add(Mod(rr=pg.Range(time(10), time(11))))

        s.flush()

        stmt = sa.select(Mod)
        mods = s.scalars(stmt).all()
        print([m.rr for m in mods])
```
errors with
```
s.flush()
...
sqlalchemy.exc.UnsupportedCompilationError: Compiler <sqlalchemy.dialects.postgresql.base.PGTypeCompiler object at 0x000001EA4132F5F0> can't render element of type TIMERANGERangeImpl() (Background on this error at: https://sqlalche.me/e/21/l7de)
```

**Reply by zzzeek** [reply]
can you confim is fails prior to 4006cb38e13ac471655f5 since that refactored all of this

**Reply by CaselIT** [reply]
it's not easy to make it work, this is an ugly patch
```diff
diff --git a/lib/sqlalchemy/dialects/postgresql/ranges.py b/lib/sqlalchemy/dialects/postgresql/ranges.py
index b793ca49f..3f739ba0e 100644
--- a/lib/sqlalchemy/dialects/postgresql/ranges.py
+++ b/lib/sqlalchemy/dialects/postgresql/ranges.py
@@ -777,10 +777,16 @@ class AbstractRange(sqltypes.TypeEngine[_T]):
             # The adapt() operation here is cached per type-class-per-dialect,
             # so is not much of a performance concern
             visit_name = self.__visit_name__
+            new_cls_dict = {"__visit_name__": visit_name}
+            if hasattr(self.__class__, "_compiler_dispatch"):
+                new_cls_dict["_compiler_dispatch"] = (
+                    self.__class__._compiler_dispatch
+                )
+            # breakpoint()
             return type(  # type: ignore
                 f"{visit_name}RangeImpl",
                 (cls, self.__class__),
-                {"__visit_name__": visit_name},
+                new_cls_dict,
             )()
         else:
             return super().adapt(cls)
```

once that's done we can make it work, but it's quite painful still since psycog knows nothing about that range so it returns strings back
```py
from datetime import time

import sqlalchemy as sa
from sqlalchemy import orm
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.dialects import postgresql as pg


class TIMERANGE_a(pg.AbstractSingleRange[time]):
    """Represent the custom PostgreSQL TIMERANGE type."""

    __visit_name__ = "TIMERANGE"


@compiles(TIMERANGE_a, "postgresql")
def compile_timerange(element, compiler, **kw):
    print(">>", type(element))
    # 1/0
    return "TIMERANGE"


class TIMERANGE(sa.TypeDecorator):
    impl = TIMERANGE_a
    cache_ok = True

    def result_processor(self, dialect, coltype):
        def to_range(value):
            if value == "empty":
                value = pg.Range(empty=True)
            elif value is not None:
                l, _, u = value[1:-1].partition(",")
                value = pg.Range(
                    time.fromisoformat(l) if l else None,
                    time.fromisoformat(u) if u else None,
                    bounds=value[0] + value[-1],
                    # empty=not value._bounds,
                )
            return value

        return to_range


e = sa.create_engine("postgresql+psycopg://scott:tiger@127.0.0.1:5432/test")
e.connect().close()


class Base(orm.DeclarativeBase):
    pass


class Mod(Base):
    __tablename__ = "mod"
    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)
    rr: orm.Mapped[pg.Range[time]] = orm.mapped_column(TIMERANGE)


e.echo = True
with e.connect() as c:
    c.exec_driver_sql("CREATE TYPE timerange AS RANGE (SUBTYPE = time)")
    Base.metadata.create_all(c)

    with orm.Session(c) as s:
        s.add(Mod(rr=pg.Range(time(10), time(11))))
        s.add(Mod(rr=pg.Range(empty=True)))
        s.add(Mod(rr=pg.Range(time(10))))
        s.add(Mod(rr=pg.Range(upper=time(10))))

        s.flush()

        stmt = sa.select(Mod)
        mods = s.scalars(stmt).all()
        print([m.rr for m in mods])
```
prints
```
[Range(lower=datetime.time(10, 0), upper=datetime.time(11, 0), bounds='[)', empty=False), Range(lower=None, upper=None, bounds='[)', empty=True), Range(lower=datetime.time(10, 0), upper=None, bounds='[)', empty=False), Range(lower=None, upper=datetime.time(10, 0), bounds='()', empty=False)]
```

**Reply by CaselIT** [reply]
> can you confim is fails prior to [4006cb3](https://github.com/sqlalchemy/sqlalchemy/commit/4006cb38e13ac471655f5f27102678ed8933ee60) since that refactored all of this

yes, the issue is not there it's in adapt. It was introduced here 9ae645d5d1a8cc7732a6d335be6205d0b21e31b1

**Reply by zzzeek** [reply]
OK we copy the visit_name here when we make the new type, we should include a fixed `_compiler_dispatch()` method that always calls down to the original type's `_compiler_dispatch()`, which should always be there

the poster can supply a hardcoded adapt() method for an interim workaround

**Reply by CaselIT** [reply]
Another issue is that this change by itself it's not very useful, since the new type will override any defined method with thes defined in the dialect impl, like if a the result_processor was defined it would be ignored and instead the one from _PsycopgRange would be used (that's not good here since psycopg returns strings here)

**Reply by TanelS** [reply]
@CaselIT and @zzzeek , thank you both for extremely quick and helpful response! Happy Easter!
